### PR TITLE
Added new container type: AspectRatioContainer

### PR DESCRIPTION
**Pull request by UgisBrekis**
This is based on the proposal(there's a video if you don't want to read the whole thing): https://github.com/godotengine/godot-proposals/issues/69

### CODE REVIEW

File: scene/gui/aspect_ratio_container.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		HEIGHT_CONTROLS_WIDTH,
+		FIT,
+		COVER,
+	};
+
+private:
+	float ratio;
+	StretchMode stretch_mode;
+	float alignment_x;
+	float alignment_y;
```
Comment (MunWolf) [inline scene/gui/aspect_ratio_container.h]: Convention would be to expose these as a `Vector2 alignment;` instead of 2 separate values I am pretty sure.

**Comment by groud** [comment]
The idea looks good. But I don't think the AlignmentX/Y are very useful, it might be worth removing them.

File: scene/gui/aspect_ratio_container.h (hunk 1 above)
Comment (UgisBrekis) [reply scene/gui/aspect_ratio_container.h]: Makes sense, I'll fix it.

**Comment by UgisBrekis** [comment]
@groud I'm gonna disagree on this. At Product Madness I've been implementing lots and lots of UI elements and I constantly use RectTransform's pivot(we use Unity) to control the alignment. In some cases it is possible to assemble UI without it, but you'll end up adding dummy Control nodes and then anchor child nodes to achieve the same effect what leads to hierarchy becoming bloated/unreadable. I actually think that other nodes, such as HBox and VBox containers, would benefit from AlignmentX/Y, since at the moment they use Begin/Center/End to control the same thing, but with limited capability. CenterContainer in general could be refactored to "AlignmentContainer" with default values set to 0.5, 0.5, so that it acts as a center container and if necessary the user can align child nodes differently.

**Comment by groud** [comment]
>  but you'll end up adding dummy Control nodes and then anchor child nodes to achieve the same effect what leads to hierarchy becoming bloated/unreadable

This is exactly why the anchors and margin workflow exists. It provides more or less the exact same feature as Unity's RectTransform does. And no, replacing a container by a basic control node does not make the hierarchy unreadable.

> CenterContainer in general could be refactored to "AlignmentContainer" with default values set to 0.5, 0.5, so that it acts as a center container and if necessary the user can align child nodes differently.

All of this is already covered by the anchor and margins workflow, there is no need to makes things more complex here.

**Comment by UgisBrekis** [comment]
> This is exactly why the anchors and margin workflow exists. It provides more or less the exact same feature as Unity's RectTransform does. And no, replacing a container by a basic control node does not make the hierarchy unreadable.

DummyNodeExample

1) Container nodes ignore child anchors, so anchor and margin workflow won't work here.
2) If it's just three nodes in a tree it is not a problem, but if you have a complex UI popup with 30 nodes, it can easily get cluttered. I understand that this can be very subjective, because some people keep things nice and clean and some just don't, but if we can provide some help with that, we should use the opportunity. 
3) It is not performance friendly to introduce dummy node to complete a simple task of alignment.
4) There are alignment options already present in HBox and VBox containers, so there is no reason to get rid of these options in AspectRatio container. All I did here is to provide more flexibility over the alignment.
5) AlignmentX/Y on other nodes is a bigger discussion and falls out of the scope of this pull request, however I don't think a float value of alignment if more complex than enum value.

**Comment by groud** [comment]
In fact, I think you don't get the philosophy behind the Containers' based workflow. We could put all possible behaviors in a single node with several options to fine tune its behavior, but that's not the philosophy behind Containers. Those nodes are meant to be kept **simple**, and you create a layout by **combining several of those simple nodes**. So it is expected that things may become clustered, and it is not a problem.

Adding options to fine tune the behavior of each Container make things harder to maintain, and causes confusion on newcomers.

Regarding the few other points:

> 3. It is not performance friendly to introduce dummy node to complete a simple task of alignment.

GUI is not a performance-critical topic at all, so it is out of question here.

> 4. There are alignment options already present in HBox and VBox containers, so there is no reason to get rid of these options in AspectRatio container. All I did here is to provide more flexibility over the alignment.

The HBoxContainer and VBoxContainer nodes are meant to contain several nodes, that's why the alignment option makes sense. When you have several nodes aligned horizontally with different height, it makes sense to be able to choose how they align vertically. And top/bottom/center are enough option, other ones have no significant use cases.

**Comment by UgisBrekis** [comment]
> The HBoxContainer and VBoxContainer nodes are meant to contain several nodes, that's why the alignment option makes sense. When you have several nodes aligned horizontally with different height, it makes sense to be able to choose how they align vertically. And top/bottom/center are enough option, other ones have no significant use cases.

I really don't understand the rationale here. Correct me if I'm wrong, you state that the user controls vertical alignment in HBoxContainer and horizontal alignment in VBoxContainer? If so, then you are not correct, that's not how it works. On HBoxContainer you control horizontal alignment, on VBoxContainer you control vertical alignment. I attached example image- it's HBoxContainer with alignments: Begin, Center, End. Based on this philosophy there is no reason to remove the ability from AspectRatioContainer to control the alignment. 

ButtonAlignment

> We could put all possible behaviors in a single node with several options to fine tune its behavior, but that's not the philosophy behind Containers. Those nodes are meant to be kept simple, and you create a layout by combining several of those simple nodes. So it is expected that things may become clustered, and it is not a problem.

I'm not suggesting to put all possible behaviours- I'm re-using existing behaviour just like I mentioned above. The only addition is float value instead of enum. And as I have stated before- you simply can't achieve the alignment solution with a dummy node. If a dummy node is positioned in top-left corner(because of the stripped ability to control alignment), your actual child object that you're trying to position will only be positioned relatively to a dummy node, not the AspectRatioContainer.

Not thinking of performance and clutter is something new, and I'm not saying it in a cynical way.

**Comment by Calinou** [comment]
> Not thinking of performance and clutter is something new, and I'm not saying it in a cynical way.

What @groud meant is that adding a few dozen dummy nodes (the worst-case scenario) probably won't cause GUI rendering to slow down below 60 FPS.

**Comment by UgisBrekis** [comment]
I'm just being cautious about this, since, for example, Unity UI is very slow. We're forced to rewrite lots of their native stuff, so that complex UI would work smoothly on very low-end devices.

**Comment by groud** [comment]
> I really don't understand the rationale here. Correct me if I'm wrong, you state that the user controls vertical alignment in HBoxContainer and horizontal alignment in VBoxContainer? If so, then you are not correct, that's not how it works. On HBoxContainer you control horizontal alignment, on VBoxContainer you control vertical alignment. I attached example image- it's HBoxContainer with alignments: Begin, Center, End. Based on this philosophy there is no reason to remove the ability from AspectRatioContainer to control the alignment.

Ah indeed you are right and I was wrong here. But in that case, I do think this option could be removed from the H/VBoxContainers.

Also, if we want it to be added, it will have to be consistent with the HBox and VBox containers, so it will have to be a enum. Consistency is more important here.

But once again, I see no use case in providing alignment here. No use case = no feature, that's it.
It could be added later on if someone really needs it, but for now, it has more drawbacks than advantages.

> Not thinking of performance and clutter is something new, and I'm not saying it in a cynical way.

Yes, as @Calinou pointed out, adding a few nodes wont hurt performances at all. Even on mobile, performances are constrained by other parts of the code. So consistency and ease of use are more important here.

**Comment by UgisBrekis** [comment]
Ok, I'll change that to enum to keep things consistent, however I would like this whole alignment topic to be brought to a bigger crowd for discussion eventually.

**Comment by groud** [comment]
> Ok, I'll change that to enum to keep things consistent, however I would like this whole alignment topic to be brought to a bigger crowd for discussion eventually.

Yeah I agree. The whole set of control should probably be someday reviewed and cleaned up. There are several inconsistencies and piled up workarounds which should likely be solved.

**Comment by aaronfranke** [comment]
> There are several inconsistencies and piled up workarounds which should likely be solved.

@groud Could you (or anyone else) make a list of them and suggest solutions, probably in https://github.com/godotengine/godot-proposals? 4.0 is coming up soon, so we could do that for 4.0.

**Comment by aaronfranke** [comment]
@UgisBrekis Is this still desired? If so, it needs to be rebased on the latest master branch.

**Comment by akien-mga** [comment]
Superseded by #43807.

**Pull request closed without merging**